grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java

```diff
@@ -431,6 +431,7 @@ private void invokeHalfClose() {
         try (SafeCloseable ignored = ctx.push()) {
             assert listener != null;
             listener.onHalfClose();
+            listener.onReady();
```

Comment:
Is there any chance where `onReady()` is called multiple times? e.g. `invokeHalfClosed()` is called after `setListener()`.

Reply:
I missed pushing my last work. (https://github.com/line/armeria/pull/3099/commits/0ff6981f8f1a4dd07d9b660a7c0b354f14ab005b)
I do not fully understand the `listener`'s life cycle. Based on the upstream gRPC server, the `listener.onReady()` is called two times if it is an unary stream listener.
1) When a stream is activated
2) `onHalfClose()` is invoked

---
-  `onHalfClose` in `StreamingServerCallListener` does not call `onReady`.
https://github.com/grpc/grpc-java/blob/9b73e2365da502a466b01544f102cd487e374428/stub/src/main/java/io/grpc/stub/ServerCalls.java#L263-L267
```java
      @Override
      public void onHalfClose() {
        halfClosed = true;
        requestObserver.onCompleted();
      }
```

- `onHalfClose` in `UnaryServerCallListener` calls `onReady`.
https://github.com/grpc/grpc-java/blob/9b73e2365da502a466b01544f102cd487e374428/stub/src/main/java/io/grpc/stub/ServerCalls.java#L170-L190
```java
      public void onHalfClose() {
        if (!canInvoke) {
          return;
        }
        if (request == null) {
          // Safe to close the call, because the application has not yet been invoked
          call.close(
              Status.INTERNAL.withDescription(MISSING_REQUEST),
              new Metadata());
          return;
        }


        method.invoke(request, responseObserver);
        request = null;
        responseObserver.freeze();
        if (wasReady) {
          // Since we are calling invoke in halfClose we have missed the onReady
          // event from the transport so recover it here.
          onReady();
        }
      }
```

Reply:
I see. So it's meant to be invoked multiple times whenever something's ready. Thanks for explanation!

Reply:
Yes. :-)

Reply:
FJYI, the Javadoc for `Listener.onReady()
```
     * This indicates that the call may now be capable of sending additional messages (via
     * {@link #sendMessage}) without requiring excessive buffering internally. This event is
     * just a suggestion and the application is free to ignore it, however doing so may
```
https://github.com/grpc/grpc-java/blob/18e7e2ddca1323bcdfec4649d7a1dd91a68f8223/api/src/main/java/io/grpc/ServerCall.java#L86-L88